File: src/bound_entangled/cd_otimes_cd/horodecki_2_by_d_generalized.py
"""https://arxiv.org/pdf/1203.3711"""

from bound_entangled.utils import ketbra
import numpy as np
from math import sqrt


def ket_ij(dims: tuple[int, int], i: int, j: int) -> np.ndarray:
    d1, d2 = dims
    ret = np.zeros(d1 * d2, dtype=np.float64)
    ret[i * d2 + j] = 1
    return ret


def ket_i(d: int, i: int) -> np.ndarray:
    ret = np.zeros(d, dtype=np.float64)
    ret[i] = 1
    return ret


def psi_i(d: int, i: int) -> np.ndarray:
    one_sqrt_2 = 1 / sqrt(2)
    return one_sqrt_2 * (ket_ij((2, d), 0, i) + ket_ij((2, d), 1, i + 1))


def rho_insep_d(d: int) -> np.ndarray:
    psi_is = 2 / (2 * d - 1) * sum([ketbra(psi_i(d, i)) for i in range(d - 1)])
    one_zero = (1 / (2 * d - 1)) * ketbra(ket_ij((2, d), 1, 0))
    return psi_is + one_zero


def phi_b(d: int, b: float) -> np.ndarray:
    zero = 1 / sqrt(2) * ket_i(2, 0)
    one_m_b_zero = sqrt(1 - b) * ket_i(d, 0)
    one_p_b_d_m_one = sqrt(1 + b) * ket_i(d, d - 1)
    return np.kron(zero, one_m_b_zero + one_p_b_d_m_one)


def horodecki_2_by_d_generalized(*, second_dim_d: int, b: float) -> np.ndarray:
    """Generalization of the C^2 ⊗ C^4 horodecki state
    for d = 4, it is equivalent to the C^2 ⊗ C^4 horodecki state from toqito up to an anti-diagonal transpose

    assert np.allclose(horodecki_2_by_d_generalized(second_dim_d=4, b=b), np.flip(horodecki(b)).T)

    Args:
        second_dim_d (int): second dimension d. The full dimension of the system is 2*d
        b (float): parameter d

    Returns:
        np.ndarray: the generalized horodecki state
    """
    return ((2 * second_dim_d - 1) * b * rho_insep_d(second_dim_d) + ketbra(phi_b(second_dim_d, b))) / ((2 * second_dim_d - 1) * b + 1)

File: js/src/cdOtimesCd/horodecki2ByDGeneralized.ts
/**
 * https://arxiv.org/pdf/1203.3711
 *
 * A generalization of the C² ⊗ C⁴ Horodecki state to C² ⊗ C^d.
 */

import { add, divide, multiply } from 'mathjs';
import type { Matrix } from 'mathjs';
import { ket, ketbra, tensor } from '../utils/index.js';
import { fail, sumMatrices } from '../utils/internal.js';

export interface Horodecki2ByDOptions {
  /** The second dimension d; the full system is C² ⊗ C^d. Must be >= 2. */
  secondDimD: number;
  /** The free real parameter b. */
  b: number;
}

/** (|0 i> + |1 i+1>) / √2, one of the d - 1 vectors of the inseparable part. */
function psiI(d: number, i: number): Matrix {
  return divide(add(ket([2, d], [0, i]), ket([2, d], [1, i + 1])), Math.SQRT2) as Matrix;
}

/** The inseparable component: a mixture of the psi_i with the corner |1 0>. */
function rhoInsep(d: number): Matrix {
  const terms: Matrix[] = [];
  for (let i = 0; i < d - 1; i++) {
    terms.push(ketbra(psiI(d, i)));
  }
  const psis = multiply(2 / (2 * d - 1), sumMatrices(terms)) as Matrix;
  const corner = multiply(1 / (2 * d - 1), ketbra(ket([2, d], [1, 0]))) as Matrix;
  return add(psis, corner);
}

/** |0>/√2 ⊗ (√(1-b)|0> + √(1+b)|d-1>), the separable component's vector. */
function phiB(d: number, b: number): Matrix {
  const qubit = multiply(1 / Math.SQRT2, ket(2, 0)) as Matrix;
  const shield = add(
    multiply(Math.sqrt(1 - b), ket(d, 0)),
    multiply(Math.sqrt(1 + b), ket(d, d - 1)),
  ) as Matrix;
  return tensor(qubit, shield);
}

/**
 * The generalization of the C² ⊗ C⁴ Horodecki state to C² ⊗ C^d.
 *
 * For d = 4 this is the C² ⊗ C⁴ Horodecki state up to an anti-diagonal
 * transpose.
 *
 * @param options - see {@link Horodecki2ByDOptions}.
 * @returns the generalized Horodecki state, of dimension 2d.
 */
export function horodecki2ByDGeneralized({ secondDimD, b }: Horodecki2ByDOptions): Matrix {
  const d = secondDimD;
  if (!Number.isInteger(d) || d < 2) {
    fail(`the second dimension must be an integer >= 2, got ${d}`);
  }
  const weight = (2 * d - 1) * b;
  return divide(
    add(multiply(weight, rhoInsep(d)) as Matrix, ketbra(phiB(d, b))),
    weight + 1,
  ) as Matrix;
}
